### PR TITLE
Feature/mirror refactoring v3

### PR DESCRIPTION
**Pull request by qqmbr4k**

### CODE REVIEW

**Comment by pferrel** [comment]
@andreyshikov what tests have been done with this branch? This is a very large PR.

File: rest-server/core/src/main/scala/com/actionml/core/backup/Mirror.scala [hunk 1]
```diff
@@ -14,53 +14,44 @@
...
     */
-  protected def containerName: String = s"$mirrorContainer${File.separator}$engineId"
+  protected val containerName: String = s"$mirrorContainer${File.separator}$engineId${File.separator}"
 
-}
 
-/** Used when mirroring is config selectable */
-object Mirror {
-  val localfs = "localfs"
-  val hdfs = "hdfs"
```
Comment (pferrel) [inline rest-server/core/src/main/scala/com/actionml/core/backup/Mirror.scala:65]: can we put these back?

File: rest-server/core/src/main/scala/com/actionml/core/engine/Engine.scala [hunk 2]
```diff
@@ -136,21 +141,39 @@ abstract class Engine extends LazyLogging with JsonSupport {
...
+    zio.Runtime.default.unsafeRunSync(mirroring.mirrorEvent(json.replace("\n", " ") + "\n"))
+    Valid(Comment("Input processed by base Engine"))
   }
 
-  def inputAsync(json: String)(implicit ec: ExecutionContext): Future[Validated[ValidateError, Response]] = {
-    Future.successful(Invalid(NotImplemented()))
-  }
+  def inputAsync(json: String)(implicit ec: ExecutionContext): Future[Validated[ValidateError, Response]] = Future(input(json))
 
   def inputMany(data: Seq[String]): Unit = {}
```
Comment (pferrel) [inline rest-server/core/src/main/scala/com/actionml/core/engine/Engine.scala]: not used?

File: rest-server/server/src/main/resources/reference.conf [hunk 3]
```diff
@@ -50,3 +44,11 @@ etcd-config {
...
   timeout = ${?HARNESS_ETCD_TIMEOUT}
 }
+
+hdfs {
+  master = "localhost"
+  master = ${?HDFS_HOST}
+  port = 9000
+  port = ${?HDFS_PORT}
+  config-dir = ${?HDFS_CONF_DIR}
+}
```
Comment (pferrel) [inline rest-server/server/src/main/resources/reference.conf]: is this a new .conf entry?

File: rest-server/server/src/main/scala/com/actionml/router/http/routes/QueriesRouter.scala [hunk 4]
```diff
@@ -62,10 +73,14 @@ class QueriesRouter(implicit inj: Injector) extends BaseRouter with Authorizatio
...
-//      (queryService ? GetPrediction(engineId, JsonMethods.compact(query))).mapTo[Validated[ValidateError, Response]]
-//    }
+    completeByValidated(StatusCodes.Created) {
+      queryService.queryAsync(engineId, JsonMethods.compact(query))
+        .map(Valid(_))
+        .recover { case e =>
+          logger.error("Prediction error", e)
+          Invalid(ValidRequestExecutionError("Prediction error"))
+        }
+    }
```
Comment (pferrel) [inline rest-server/server/src/main/scala/com/actionml/router/http/routes/QueriesRouter.scala:78]: refactor?

**Comment by pferrel** [comment]
@andreyshikov have you tested `hctl cancel` for imports?

**Comment by pferrel** [comment]
@andreyshikov have you tested bad config with mirror location set to hdfs but no hdfs config setup in harness-env ?

**Comment by pferrel** [comment]
Tests to run:
 - import from file
 - mirror live event input
 - mirror as an import is happening, this should add newly imported events to the mirror exactly as they are imported just like live event mirroring
 - import from mirrored location while it is setup as a mirror -- IF this is not an error it will recursively import from newly mirrored events infinitely

You may have tested these but we need them all verified/tested

**Comment by andreyshikov** [comment]
> @andreyshikov have you tested `hctl cancel` for imports?

yes, cancel works for imports

**Comment by andreyshikov** [comment]
> @andreyshikov have you tested bad config with mirror location set to hdfs but no hdfs config setup in harness-env ?

neither import with wrong hdfs uri nor adding of engine with a bad hdfs config can't be done. hctl will get an error

**Comment by pferrel** [comment]
@andreyshikov if all the use case tests work except the infinite loop import. Go ahead and add an issue report then merge.

**Pull request merged**